File: lib/tools/serpapi.ts
import { getCached, setCache } from '../supabase';
import type { ToolResult, SearchResult, TrendPoint } from './types';
import { buildToolResult } from './fallback';
import { getSearxngBaseUrl, searchSearxng } from './searxng';

const BASE_URL = 'https://serpapi.com/search';

/** Optional locale for geo-grounded search (SerpAPI gl/hl + SearXNG language). */
export type SearchLocaleOptions = {
  gl?: string;       // country code, e.g. "lk"
  hl?: string;       // language, e.g. "en"
  location?: string; // free-text location bias
};

class SerpError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SerpError';
  }
}

async function serpFetch(params: Record<string, string>): Promise<unknown> {
  const apiKey = process.env.SERPAPI_KEY;
  if (!apiKey) throw new SerpError('SERPAPI_KEY not set');
  const url = new URL(BASE_URL);
  url.searchParams.set('api_key', apiKey);
  for (const [k, v] of Object.entries(params)) {
    url.searchParams.set(k, v);
  }
  const res = await fetch(url.toString(), {
    next: { revalidate: 0 },
    signal: AbortSignal.timeout(15000),
  });
  if (!res.ok) throw new SerpError(`SerpAPI ${res.status}: ${await res.text()}`);
  return res.json();
}

function emptySearchResult(source: string, query: string): ToolResult<SearchResult[]> {
  return buildToolResult<SearchResult[]>({
    data: [],
    status: 'failed',
    source: `${source} (failed)`,
    sourceUrl: `https://google.com/search?q=${encodeURIComponent(query)}`,
  });
}

function hasSerpApiKey(): boolean {
  return Boolean(process.env.SERPAPI_KEY?.trim());
}

function localeCacheSuffix(locale?: SearchLocaleOptions): string {
  if (!locale) return '';
  const parts = [locale.gl, locale.hl, locale.location].filter(Boolean);
  return parts.length ? `:loc:${parts.join(',')}` : '';
}

function applyLocaleParams(
  base: Record<string, string>,
  locale?: SearchLocaleOptions,
): Record<string, string> {
  if (!locale) return base;
  const out = { ...base };
  if (locale.gl?.trim()) out.gl = locale.gl.trim().toLowerCase();
  if (locale.hl?.trim()) out.hl = locale.hl.trim().toLowerCase();
  if (locale.location?.trim()) out.location = locale.location.trim();
  return out;
}

/**
 * Prefer SearXNG when configured; fall back to SerpAPI Google organic.
 */
async function searchWebViaSerpApi(
  query: string,
  locale?: SearchLocaleOptions,
): Promise<ToolResult<SearchResult[]>> {
  try {
    const raw = (await serpFetch(
      applyLocaleParams({ engine: 'google', q: query, num: '10' }, locale),
    )) as {
      organic_results?: Array<{ title: string; link: string; snippet?: string; date?: string }>;
    };

    const results: SearchResult[] = (raw.organic_results ?? []).slice(0, 8).map(r => ({
      title: r.title,
      url: r.link,
      snippet: r.snippet ?? '',
      date: r.date,
    }));

    return buildToolResult<SearchResult[]>({
      data: results,
      status: results.length > 0 ? 'ok' : 'failed',
      source: 'SerpAPI / Google',
      sourceUrl: `https://google.com/search?q=${encodeURIComponent(query)}`,
    });
  } catch {
    return emptySearchResult('SerpAPI / Google', query);
  }
}

async function searchNewsViaSerpApi(
  query: string,
  locale?: SearchLocaleOptions,
): Promise<ToolResult<SearchResult[]>> {
  try {
    const raw = (await serpFetch(
      applyLocaleParams(
        {
          engine: 'google',
          q: query,
          tbm: 'nws',
          num: '10',
        },
        locale,
      ),
    )) as {
      news_results?: Array<{ title: string; link: string; snippet?: string; date?: string }>;
    };

    const results: SearchResult[] = (raw.news_results ?? []).slice(0, 8).map(r => ({
      title: r.title,
      url: r.link,
      snippet: r.snippet ?? '',
      date: r.date,
    }));

    return buildToolResult<SearchResult[]>({
      data: results,
      status: results.length > 0 ? 'ok' : 'failed',
      source: 'SerpAPI / Google News',
      sourceUrl: `https://news.google.com/search?q=${encodeURIComponent(query)}`,
    });
  } catch {
    return emptySearchResult('SerpAPI / Google News', query);
  }
}

export async function searchWeb(
  query: string,
  locale?: SearchLocaleOptions,
): Promise<ToolResult<SearchResult[]>> {
  const cacheKey = `web:${query}${localeCacheSuffix(locale)}`;
  const cached = await getCached('serpapi_search', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<SearchResult[]>), cached: true };
  }

  const searxLanguage = locale?.hl?.trim() || undefined;

  if (getSearxngBaseUrl()) {
    const searx = await searchSearxng(query, {
      categories: 'general',
      language: searxLanguage,
    });
    if (searx.data.length > 0 && searx.status !== 'failed') {
      await setCache('serpapi_search', cacheKey, searx);
      return searx;
    }
    if (hasSerpApiKey()) {
      const serp = await searchWebViaSerpApi(query, locale);
      if (serp.data.length > 0) {
        const degraded = buildToolResult<SearchResult[]>({
          data: serp.data,
          status: 'degraded',
          source: `${serp.source} (SearXNG miss)`,
          sourceUrl: serp.sourceUrl,
        });
        await setCache('serpapi_search', cacheKey, degraded);
        return degraded;
      }
      await setCache('serpapi_search', cacheKey, serp);
      return serp;
    }
    await setCache('serpapi_search', cacheKey, searx);
    return searx;
  }

  if (!hasSerpApiKey()) {
    return emptySearchResult('SerpAPI / Google', query);
  }

  const result = await searchWebViaSerpApi(query, locale);
  await setCache('serpapi_search', cacheKey, result);
  return result;
}

export async function searchNews(
  query: string,
  locale?: SearchLocaleOptions,
): Promise<ToolResult<SearchResult[]>> {
  const cacheKey = `news:${query}${localeCacheSuffix(locale)}`;
  const cached = await getCached('serpapi_news', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<SearchResult[]>), cached: true };
  }

  const searxLanguage = locale?.hl?.trim() || undefined;

  if (getSearxngBaseUrl()) {
    const searx = await searchSearxng(query, {
      categories: 'news',
      language: searxLanguage,
    });
    if (searx.data.length > 0 && searx.status !== 'failed') {
      await setCache('serpapi_news', cacheKey, searx);
      return searx;
    }
    if (hasSerpApiKey()) {
      const serp = await searchNewsViaSerpApi(query, locale);
      if (serp.data.length > 0) {
        const degraded = buildToolResult<SearchResult[]>({
          data: serp.data,
          status: 'degraded',
          source: `${serp.source} (SearXNG miss)`,
          sourceUrl: serp.sourceUrl,
        });
        await setCache('serpapi_news', cacheKey, degraded);
        return degraded;
      }
      await setCache('serpapi_news', cacheKey, serp);
      return serp;
    }
    await setCache('serpapi_news', cacheKey, searx);
    return searx;
  }

  if (!hasSerpApiKey()) {
    return emptySearchResult('SerpAPI / Google News', query);
  }

  const result = await searchNewsViaSerpApi(query, locale);
  await setCache('serpapi_news', cacheKey, result);
  return result;
}

export async function searchTrends(keywords: string[]): Promise<ToolResult<TrendPoint[]>> {
  const cacheKey = `trends:${keywords.join(',')}`;
  const cached = await getCached('serpapi_trends', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<TrendPoint[]>), cached: true };
  }

  try {
    const raw = (await serpFetch({
      engine: 'google_trends',
      q: keywords.join(','),
      data_type: 'TIMESERIES',
      date: 'today 12-m',
    })) as {
      interest_over_time?: {
        timeline_data?: Array<{
          date: string;
          values?: Array<{ value?: string; query: string }>;
        }>;
      };
    };

    const points: TrendPoint[] = [];
    const timeline = raw.interest_over_time?.timeline_data ?? [];

    for (const point of timeline) {
      for (const kw of point.values ?? []) {
        points.push({
          date: point.date,
          value: Number.parseInt(kw.value ?? '0', 10),
          keyword: kw.query,
        });
      }
    }

    const result = buildToolResult<TrendPoint[]>({
      data: points,
      status: points.length > 0 ? 'ok' : 'failed',
      source: 'SerpAPI / Google Trends',
      sourceUrl: `https://trends.google.com/trends/explore?q=${encodeURIComponent(keywords.join(','))}`,
    });

    await setCache('serpapi_trends', cacheKey, result);
    return result;
  } catch {
    return buildToolResult<TrendPoint[]>({
      data: [],
      status: 'failed',
      source: 'SerpAPI / Google Trends (failed)',
      sourceUrl: `https://trends.google.com/trends/explore?q=${encodeURIComponent(keywords.join(','))}`,
    });
  }
}

export async function searchAdsTransparency(advertiser: string): Promise<ToolResult<SearchResult[]>> {
  const cacheKey = `ads:${advertiser}`;
  const cached = await getCached('serpapi_search', cacheKey);
  if (cached) {
    return { ...(cached as ToolResult<SearchResult[]>), cached: true };
  }

  try {
    const raw = (await serpFetch({
      engine: 'google',
      q: `"${advertiser}" site:adstransparency.google.com OR "${advertiser}" ads`,
      num: '5',
    })) as {
      organic_results?: Array<{ title: string; link: string; snippet?: string }>;
    };

    const results: SearchResult[] = (raw.organic_results ?? []).slice(0, 5).map(r => ({
      title: r.title,
      url: r.link,
      snippet: r.snippet ?? '',
    }));

    const result = buildToolResult<SearchResult[]>({
      data: results,
      status: results.length > 0 ? 'ok' : 'failed',
      source: 'SerpAPI / Google Ads Transparency',
    });

    await setCache('serpapi_search', cacheKey, result);
    return result;
  } catch {
    return emptySearchResult('SerpAPI / Google Ads Transparency', advertiser);
  }
}

File: backend/app/tools/serpapi.py
"""SerpAPI tools — mirrors lib/tools/serpapi.ts."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from app.config import get_settings
from app.supabase_client import get_cached, set_cache

from .fallback import build_tool_result

BASE_URL = "https://serpapi.com/search"
TIMEOUT = 15.0


async def _serp_fetch(params: dict[str, str]) -> dict[str, Any]:
    settings = get_settings()
    if not settings.serpapi_key:
        raise RuntimeError("SERPAPI_KEY not set")

    query_params = {"api_key": settings.serpapi_key, **params}
    async with httpx.AsyncClient(timeout=TIMEOUT) as client:
        res = await client.get(BASE_URL, params=query_params)
        if not res.is_success:
            raise RuntimeError(f"SerpAPI {res.status_code}: {res.text}")
        return res.json()


def _empty_search_result(source: str, query: str) -> dict[str, Any]:
    return build_tool_result(
        data=[],
        status="failed",
        source=f"{source} (failed)",
        source_url=f"https://google.com/search?q={quote(query)}",
    )


async def search_web(query: str) -> dict[str, Any]:
    cache_key = f"web:{query}"
    cached = await get_cached("serpapi_search", cache_key)
    if cached:
        return {**cached, "cached": True}

    try:
        raw = await _serp_fetch({"engine": "google", "q": query, "num": "10"})
        results = [
            {
                "title": r.get("title", ""),
                "url": r.get("link", ""),
                "snippet": r.get("snippet") or "",
                "date": r.get("date"),
            }
            for r in (raw.get("organic_results") or [])[:8]
        ]
        result = build_tool_result(
            data=results,
            status="ok" if results else "failed",
            source="SerpAPI / Google",
            source_url=f"https://google.com/search?q={quote(query)}",
        )
        await set_cache("serpapi_search", cache_key, result)
        return result
    except Exception:
        return _empty_search_result("SerpAPI / Google", query)


async def search_news(query: str) -> dict[str, Any]:
    cache_key = f"news:{query}"
    cached = await get_cached("serpapi_news", cache_key)
    if cached:
        return {**cached, "cached": True}

    try:
        raw = await _serp_fetch({"engine": "google", "q": query, "tbm": "nws", "num": "10"})
        results = [
            {
                "title": r.get("title", ""),
                "url": r.get("link", ""),
                "snippet": r.get("snippet") or "",
                "date": r.get("date"),
            }
            for r in (raw.get("news_results") or [])[:8]
        ]
        result = build_tool_result(
            data=results,
            status="ok" if results else "failed",
            source="SerpAPI / Google News",
            source_url=f"https://news.google.com/search?q={quote(query)}",
        )
        await set_cache("serpapi_news", cache_key, result)
        return result
    except Exception:
        return _empty_search_result("SerpAPI / Google News", query)


async def search_trends(keywords: list[str]) -> dict[str, Any]:
    joined = ",".join(keywords)
    cache_key = f"trends:{joined}"
    cached = await get_cached("serpapi_trends", cache_key)
    if cached:
        return {**cached, "cached": True}

    explore_url = f"https://trends.google.com/trends/explore?q={quote(joined)}"
    try:
        raw = await _serp_fetch(
            {
                "engine": "google_trends",
                "q": joined,
                "data_type": "TIMESERIES",
                "date": "today 12-m",
            }
        )
        points: list[dict[str, Any]] = []
        for point in raw.get("interest_over_time", {}).get("timeline_data") or []:
            for kw in point.get("values") or []:
                points.append(
                    {
                        "date": point.get("date", ""),
                        "value": int(kw.get("value") or 0),
                        "keyword": kw.get("query", ""),
                    }
                )
        result = build_tool_result(
            data=points,
            status="ok" if points else "failed",
            source="SerpAPI / Google Trends",
            source_url=explore_url,
        )
        await set_cache("serpapi_trends", cache_key, result)
        return result
    except Exception:
        return build_tool_result(
            data=[],
            status="failed",
            source="SerpAPI / Google Trends (failed)",
            source_url=explore_url,
        )


async def search_ads_transparency(advertiser: str) -> dict[str, Any]:
    cache_key = f"ads:{advertiser}"
    cached = await get_cached("serpapi_search", cache_key)
    if cached:
        return {**cached, "cached": True}

    try:
        raw = await _serp_fetch(
            {
                "engine": "google",
                "q": f'"{advertiser}" site:adstransparency.google.com OR "{advertiser}" ads',
                "num": "5",
            }
        )
        results = [
            {
                "title": r.get("title", ""),
                "url": r.get("link", ""),
                "snippet": r.get("snippet") or "",
            }
            for r in (raw.get("organic_results") or [])[:5]
        ]
        result = build_tool_result(
            data=results,
            status="ok" if results else "failed",
            source="SerpAPI / Google Ads Transparency",
        )
        await set_cache("serpapi_search", cache_key, result)
        return result
    except Exception:
        return _empty_search_result("SerpAPI / Google Ads Transparency", advertiser)
